polkadot/node/core/bitfield-signing/src/lib.rs

```diff
@@ -69,35 +75,115 @@ pub enum Error {
 	MpscSend(#[from] mpsc::SendError),
 
 	#[error(transparent)]
-	Runtime(#[from] RuntimeApiError),
+	Runtime(#[from] util::runtime::Error),
 
 	#[error("Keystore failed: {0:?}")]
 	Keystore(KeystoreError),
+
+	#[error("Cannot find block number for given relay parent")]
+	BlockNumberNotFound,
+
+	#[error("Retrieving response from Chain API unexpectedly failed with error: {0}")]
+	ChainApi(#[from] ChainApiError),
 }
 
 /// If there is a candidate pending availability, query the Availability Store
 /// for whether we have the availability chunk for our validator index.
 async fn get_core_availability(
 	core: &CoreState,
-	validator_idx: ValidatorIndex,
-	sender: &Mutex<&mut impl SubsystemSender<overseer::BitfieldSigningOutgoingMessages>>,
+	n_validators: usize,
+	validator_index: ValidatorIndex,
+	sender: &Mutex<&mut impl overseer::BitfieldSigningSenderTrait>,
 	span: &jaeger::Span,
 ) -> Result<bool, Error> {
 	if let CoreState::Occupied(core) = core {
 		let _span = span.child("query-chunk-availability");
+		let relay_parent = core.candidate_descriptor.relay_parent;
+
+		let block_number = get_block_number::<_, Error>(*sender.lock().await, relay_parent).await;
+
+		let block_number = match block_number {
+			Ok(Some(block_number)) => block_number,
+			Ok(None) => {
+				gum::warn!(
+					target: LOG_TARGET,
+					?relay_parent,
+					?core.candidate_hash,
+					para_id = %core.para_id(),
+					"Failed to get block number."
+				);
+
+				return Ok(false)
+			},
+			Err(err) => {
+				gum::warn!(
+					target: LOG_TARGET,
+					?relay_parent,
+					?core.candidate_hash,
+					para_id = %core.para_id(),
+					error = ?err,
+					"Failed to get block number."
+				);
+
+				return Ok(false)
+			},
+		};
+
+		let maybe_client_features = request_client_features(relay_parent, *sender.lock().await)
+			.await
+			.map_err(Error::from)?;
+
+		// Init this to all zeros. It won't be used unless
+		// `ClientFeatures::AVAILABILITY_CHUNK_SHUFFLING` is enabled. We do this to avoid querying
+		// the runtime API for session index and session info unless the feature is enabled.
+		let mut babe_randomness = [0; 32];
```

Comment:
I rely on the knowledge that if the `AVAILABILITY_CHUNK_SHUFFLING` feature is not present, the `availability_chunk_index()` function will just return the validator_index back (because prior to this feature, chunk index was always the same as the validator index) and will not use the babe_randomness, so we save one runtime API call.

Reply:
no longer relevant with latest algorithm from RFC